### PR TITLE
K8s: Add basic query service

### PR DESCRIPTION
**Pull request by ryantxu**
Check:

Plugin discovery -- eventually from api discovery:
http://localhost:3000/apis/query.grafana.app/v0alpha1/datasourceapiservers

Note that the swagger UI has some example that can work 🎉 

TODO (but in following PRs?)
* support expressions
* improve/audit tracking+instrumentation
* execute testdata queries using http client, not the plugins client
* improve the OpenAPI spec/support for backend.QueryResponseData
* restructure /ds/query to use the same (direct) runner
* and much much more

### CODE REVIEW

File: pkg/registry/apis/datasource/query.go
Comment (ryantxu) [inline pkg/registry/apis/datasource/query.go:1]: query object moved to the query service

File: pkg/apis/query/v0alpha1/query.go [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+	UID string `json:"uid"`
+}
+
+type GenericDataQuery struct {
+	// RefID is the unique identifier of the query, set by the frontend call.
+	RefID string `json:"refId"`
+
+	// TimeRange represents the query range
+	// NOTE: unlike generic /ds/query, we can now send explicit time values in each query
+	TimeRange *TimeRange `json:"timeRange,omitempty"`
```
Comment (ryantxu) [inline pkg/apis/query/v0alpha1/query.go:64]: ^^ this is new.   Currently it exists on raw backend queries, but no in the /ds/query endpoint

File: pkg/apis/query/v0alpha1/query.go [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+		}
+		delete(vals, key)
+	}
+
+	key = "datasource"
+	v, ok = vals[key]
+	if ok {
+		wrap, ok := v.(map[string]any)
+		if !ok {
+			return fmt.Errorf("expected datsource as object (got: %t)", v)
```
Comment (marefr) [inline pkg/apis/query/v0alpha1/query.go:170]: Nit. Legacy queries might have `"datasource": "<datasource name>"`?

File: pkg/api/api.go [hunk 3]
```diff
@@ -515,7 +519,22 @@ func (hs *HTTPServer) registerRoutes() {
...
+				user, err := appcontext.User(r.Context())
+				if err != nil || user == nil {
+					w.Write([]byte("no user... ???"))
+					return
+				}
+				r.URL.Path = "/apis/query.grafana.app/v0alpha1/namespaces/" + namespaceMapper(user.OrgID) + "/query"
+				hs.clientConfigProvider.DirectlyServeHTTP(w, r)
+			}
+		}
+		apiRoute.Post("/ds/query", requestmeta.SetSLOGroup(requestmeta.SLOGroupHighSlow), authorize(ac.EvalPermission(datasources.ActionQuery)), dsQueryHandler)
```
Comment (marefr) [inline pkg/api/api.go:537]: Nit. Rather than implementing this here in api.go might want to do it in metrics.go instead?

File: pkg/registry/apis/query/runner/direct.go [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+			// },
+		}
+
+		results.Items = append(results.Items, ds)
+	}
+
+	return results, nil
+}
+
+// GetDatasourcePlugins no namespace? everything that is available
```
Comment (marefr) [inline pkg/registry/apis/query/runner/direct.go:161]: It's a bit unclear what you're trying to achieve, any datasource api server available or datasource backend plugin installed per tenant?

File: pkg/tsdb/legacydata/conversions.go [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/grafana/grafana-plugin-sdk-go/backend"
+
+	"github.com/grafana/grafana/pkg/apis/query/v0alpha1"
+)
+
+// ToDataSourceQueries returns queries that should be sent to a single datasource
```
Comment (marefr) [inline pkg/tsdb/legacydata/conversions.go:13]: Nit. I would put these helpers under apis somewhere. legacydata package here refers to legacy query contracts used before the SDK contracts was introduced.

**Review by marefr — COMMENTED** [review]
Looks interesting/promising. A bit hard to follow in a few places. 

I'm a bit concerned though with duplication of existing code/logic for parsing of query requests. In a perfect world it would be nice sharing as much code as possible between existing query service and this one, but maybe not that easy perhaps?

File: pkg/api/api.go (hunk 3 above)
Comment (ryantxu) [reply pkg/api/api.go:537]: NOTE: this is likely the last thing we should do -- but is also the easiest way to check if thigns work :)

Long term metrics.go should disappear, and everything it does gets replaced by a query service.  This flag jumps the gun since obviously LOTS missing from the query service still

File: pkg/registry/apis/query/runner/direct.go (hunk 4 above)
Comment (ryantxu) [reply pkg/registry/apis/query/runner/direct.go:161]: In the "direct" mode, this seems pretty silly -- i mean, just use the wire services!

However -- the value of this approach is that the api-server is not able to decide how to resolve each query.  It may be local/multi-tenant or delegate to a single tenant grafana instance.

File: pkg/tsdb/legacydata/conversions.go (hunk 5 above)
Comment (ryantxu) [reply pkg/tsdb/legacydata/conversions.go:13]: I tried... but given the dependency on `legacydata.DataTimeRange` it seems ok as a first pass.

Perhaps with this effort we can see what parts of `legacydata` sill need to exist?

File: pkg/apis/query/v0alpha1/register.go [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+	APIVERSION = GROUP + "/" + VERSION
+)
+
+var DataSourceResourceInfo = common.NewResourceInfo(GROUP, VERSION,
+	"datasources", "datasource", "DataSource",
+	func() runtime.Object { return &DataSource{} },
+	func() runtime.Object { return &DataSourceList{} },
+)
+
+var DataSourcePluginResourceInfo = common.NewResourceInfo(GROUP, VERSION,
```
[the remark this answers is not in the document]
Comment (ryantxu) [reply pkg/apis/query/v0alpha1/register.go:22]: not needed now -- but I think this approach solves the desire need to have a single place to list datasources across all types.  This could be the API we use when listing what datasources exist for a tenant/user/token.

NOTE: we currently do not have an API that tells you your instances, it is rendered into the startup HTML and also available from `/api/frontend/settings#datasources`

File: pkg/apis/query/v0alpha1/query.go (hunk 2 above)
Comment (ryantxu) [reply pkg/apis/query/v0alpha1/query.go:170]: good point -- i'll add some more examples and tests to the parser

**Comment by ryantxu** [comment]
> I'm a bit concerned though with duplication

💯 YES -- the challenge is how to get there :)  In this pass I am trying to get query objects that support both existing and new requirements, and next we can ideally remove/replace https://github.com/grafana/grafana/blob/v10.3.1/pkg/api/dtos/models.go#L63
with the query model defined here (it is the same thing, but not `[]*simplejson.Json`)

then we can wire: https://github.com/grafana/grafana/blob/v10.3.1/pkg/services/query/query.go#L64C83-L64C102
to use this parsing strategy.

**Review by wbrowne — COMMENTED** [review]
Took it for a spin and it looks good! Just have a few comments/questions.

I assume a lot of this is copied so no need (at this stage) to go too deep into the query bits other than single datasource query?

File: pkg/registry/apis/datasource/sub_query.go [hunk 7]
```diff
@@ -77,21 +80,27 @@ func (r *subQueryREST) Connect(ctx context.Context, name string, opts runtime.Ob
...
 			PluginContext: pluginCtx,
 			Queries:       queries,
 			//  Headers: // from context
 		})
 		if err != nil {
+			responder.Error(err)
 			return
 		}
 
+		// TODO: get format (maybe protobuf!!!)
```
[the remark this answers is not in the document]
Comment (ryantxu) [reply pkg/registry/apis/datasource/sub_query.go:103]: TBD -- for machine-machine communication, the protobuf format may be a better option than JSON.  

For the grpc based plugins, it will absolutely be better/more efficient if we can just pass through the raw protobuf :) rather than converting it from `protobuf > arrow > go structs > json`

File: pkg/apis/query/v0alpha1/types.go [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// The display name
+	Title string `json:"title"`
+
+	// API Group
+	Group string `json:"group"`
+
+	// Health check (run periodically after requests?)
+	Health *HealthCheck `json:"health,omitempty"`
```
[the remark this answers is not in the document]
Comment (ryantxu) [reply pkg/apis/query/v0alpha1/types.go]: yes... removed it and we can discuss later :)

File: pkg/registry/apis/datasource/sub_query.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (ryantxu) [reply pkg/registry/apis/datasource/sub_query.go]: I added a much bigger comment to explain why I think it is important

**Comment by ryantxu** [comment]
@wbrowne @marefr -- I removed a bunch of the things we do not need now and did more cleanup.  Thoughts?  Ready to merge and tackle the followups in smaller PRs?

File: pkg/apis/query/v0alpha1/results.go [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+	spec "k8s.io/kube-openapi/pkg/validation/spec"
+)
+
+// Wraps backend.QueryDataResponse, however it includes TypeMeta and implements runtime.Object
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type QueryDataResponse struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// Backend wrapper (external dependency)
+	backend.QueryDataResponse
```
Comment (ryantxu) [inline pkg/apis/query/v0alpha1/results.go:18]: ^^^ @toddtreece -- I don't think we can avoid this, but also not sure the best approach to solve things.

I think this is a reasonable first step, but we need to find a better story for how the plugins SDK formats show up in k8s.  

Yes the openapi is a hack, but it makes everything happy.  

We can implement DeepCopy() directly in the SDK without dependency explosion, but I am not sure about TypeMeta and runtime.Object 😢

File: pkg/registry/apis/query/runner/direct.go (hunk 4 above)
Comment (ryantxu) [reply pkg/registry/apis/query/runner/direct.go:119]: I renamed this to `GetDatasourceApiServers` -- hopefully more clear

File: pkg/api/api.go [hunk 10]
```diff
@@ -515,7 +521,22 @@ func (hs *HTTPServer) registerRoutes() {
...
+				user, err := appcontext.User(r.Context())
+				if err != nil || user == nil {
+					errhttp.Write(r.Context(), fmt.Errorf("no user"), w)
+					return
+				}
+				r.URL.Path = "/apis/query.grafana.app/v0alpha1/namespaces/" + namespaceMapper(user.OrgID) + "/query"
+				hs.clientConfigProvider.DirectlyServeHTTP(w, r)
+			}
+		}
+		apiRoute.Post("/ds/query", requestmeta.SetSLOGroup(requestmeta.SLOGroupHighSlow), authorize(ac.EvalPermission(datasources.ActionQuery)), dsQueryHandler)
```
Comment (marefr) [inline pkg/api/api.go:539]: Just meant that we don't normally put any handler code/implementation  in this file api.go why it could move to metrics.go for now?

**Review by marefr — APPROVED** [review]
LGTM we can merge and tackle the followups in smaller PRs

File: pkg/api/api.go (hunk 10 above)
Comment (ryantxu) [reply pkg/api/api.go:539]: I'm not sure how to handle the `routing.Wrap(...)` part 🤔  -- the replacement is unable to return a `response.Response`, it is simply a url rewrite

Comment (marefr) [reply pkg/api/api.go:539]: Good point/question :smile: Maybe something like below, otherwise keep as is for now and we can figure out later

```go
// api.go
apiRoute.Post("/ds/query", requestmeta.SetSLOGroup(requestmeta.SLOGroupHighSlow), authorize(ac.EvalPermission(datasources.ActionQuery)), hs.DSQueryEndpoint)

// metrics.go
func (hs *HTTPServer) DSQueryEndpoint() web.Handler {
	if hs.Features.IsEnabledGlobally(featuremgmt.FlagKubernetesQueryServiceRewrite) {
		// DEV ONLY FEATURE FLAG!
		// rewrite requests from /ds/query to the new query service
		namespaceMapper := request.GetNamespaceMapper(hs.Cfg)
		return func(w http.ResponseWriter, r *http.Request) {
			user, err := appcontext.User(r.Context())
			if err != nil || user == nil {
				errhttp.Write(r.Context(), fmt.Errorf("no user"), w)
				return
			}
			r.URL.Path = "/apis/query.grafana.app/v0alpha1/namespaces/" + namespaceMapper(user.OrgID) + "/query"
			hs.clientConfigProvider.DirectlyServeHTTP(w, r)
		}
	}

	return routing.Wrap(hs.QueryMetricsV2)
}
```

**Review by wbrowne — APPROVED** [review]
Nice one 🚀
